draft-irtf-cfrg-hpke.md

```diff
@@ -296,6 +299,89 @@ plaintexts.
 The procedures described in this session are laid out in a
 Python-like pseudocode.  The ciphersuite in use is left implicit.
 
+## Key Schedule
+
+The variants of HPKE defined in this document share a common
+mechanism for translating the inputs to the protocol into an
+encryption context.  The inputs to the key schedule are as follows:
+
+* `pkR` - The receiver's public key
+* `zz` - A shared secret generated via the KEM for this transaction
+* `enc` - An encapsulated key produced by the KEM for the receiver
+* `info` (optional) - Application-supplied information
+* `psk` (optional) - A pre-shared secret held by both the initiator
+  and the receiver.
+* `pskID` (optional) - An identifier for the PSK
+* `pkI` (optional) - The initiator's public key
+
+If any of the optional values is not present (e.g., set to `None`
+in a Python implementation), then it is treated in the same way as
+an empty octet string.
+
+The key and nonce produced by this algorithm have the property that
+they are only known to the holder of the receipient private key, and
+the party that ran the KEM to generate `zz` and `enc`.  If the `psk`
+and `pskID` arguments are provided, then the recipient is assured
+that the initiator held the PSK.  If the `pkIm` argument is
+provided, then the recipient is assued that the initator held the
+corresponding private key (assuming that `zz` and `enc` were
+generated using the AuthEncap / AuthDecap methods; see below).
+
+~~~~~
+default_pkIm = zero(Npk)
+default_salt = zero(Nh)
+default_pskId = zero(0)
+
+def VerifyMode(mode, salt, pskID, pkIm):
+  got_psk = (salt != default_salt and pskID != default_pskID)
+  no_psk = (salt == default_salt and pskID == default_pskID)
+  got_pkIm = (pkIm != default_pkIm)
+  no_pkIm = (pkIm == default_pkIm)
+
+  if mode == mode_base and (got_psk or got_pkIm):
+    raise Exception("Invalid configuration for mode_base")
+  if mode == mode_psk and (no_psk or got_pkIm):
+    raise Exception("Invalid configuration for mode_psk")
+  if mode == mode_auth and (got_psk or no_pkIm):
+    raise Exception("Invalid configuration for mode_auth")
+  if mode == mode_psk_auth and (no_psk or no_pkIm):
+    raise Exception("Invalid configuration for mode_psk_auth")
+
+def KeySchedule(mode, pkRm, zz, enc, info, salt, pskID, pkIm):
+  VerifyMode(mode, psk, pskID, pkI)
+
+  pkRm = Marshal(pkR)
+  context = concat(mode, ciphersuite,
+                   len(enc), enc, len(pkRm), pkRm, pkIm,
```

Comment:
Question: if `Nenc` is fixed for the cipher suite, do we need to encode the length here? (This doesn't seem like it'd match the TLS struct below, since `opaque enc[Nenc];` is a fixed-length array.

Reply:
Good point.  Removed this length, and the length for `pkRm`, which is unnecessary for similar reasons.